R/cnd-entrace.R

```diff
@@ -1,7 +1,85 @@
+#' Entrace unexpected errors
+#'
+#' @description
+#' `global_entrace()` enriches unexpected errors (technically,
+#' unhandled errors) with rlang features:
+#'
+#' - They are assigned a backtrace.
+#'
+#' - They are recorded in [last_error()].
+#'
+#' - You can configure whether to display a backtrace on error with
+#'   the [rlang_backtrace_on_error] global option.
+#'
+#' When global entracing is enabled, all errors behave as if they had
+#' been thrown with [rlang::abort()], even the ones thrown with
+#' [stop()] or from native code.
+#'
+#' @param enable Whether to enable or disable entracing.
+#' @param type What kind of conditions should be entraced.
+#'
+#' @section Under the hood:
+#' On R 4.0 and newer, `global_entrace()` installs a global handler
+#' with `globalCallingHandlers()`. On older R versions, `entrace()` is
+#' set as an `option(error = )` handler. The latter method has the
+#' disadvantage that only one handler can be set at a time. This means
+#' that you need to manually switch between `entrace()` and other
+#' handlers like [recover()]. Also this causes a conflict with IDE
+#' handlers (e.g. in RStudio).
+#'
+#' @examples
+#' # Add this to your RProfile
+#' global_entrace()
+#'
+#' # Disable entracing
+#' global_entrace(FALSE)
+#' @export
+global_entrace <- function(enable = TRUE, type = "error") {
+  # TODO: Support "warning" and "message". Use `arg_match(multiple = TRUE)`?
+  if (!is_string(type, "error")) {
+    abort(sprintf("%s must be \"error\"", format_arg("type")))
+  }
+
+  if (getRversion() < "4.0") {
+    return(global_entrace_fallback(enable, type))
+  }
+
+  if (enable) {
+    poke_handlers <- globalCallingHandlers
+  } else {
+    poke_handlers <- drop_global_handlers
+  }
+
+  handlers <- rep_named(type, list(entrace))
+  inject(poke_handlers(!!!handlers))
+
+  invisible(NULL)
+}
+global_entrace_fallback <- function(enable, type) {
+  if (!"error" %in% type) {
+    return(invisible(NULL))
```

Comment:
I'm anticipating that the default will be `c("error", "warning", "message")` so I think it's better to ignore to avoid startup errors when we switch back to an older R.

Reply:
Ok, that makes sense. Will just need to document it.